persist/fs/seek.go

```diff
@@ -320,63 +339,53 @@ func (s *seeker) SeekByID(id ts.ID) (checked.Bytes, error) {
 // instead of looking it up on its own. Useful in cases where you've already
 // obtained an entry and don't want to waste resources looking it up again.
 func (s *seeker) SeekByIndexEntry(entry IndexEntry) (checked.Bytes, error) {
-	_, err := s.dataFd.Seek(entry.Offset, 0)
-	if err != nil {
-		return nil, err
+	// Should never happen, but prevent panics if somehow we're provided an index entry
+	// with a negative or too large offset
+	if int(entry.Offset) > len(s.dataMmap)-1 {
+		return nil, errInvalidDataFileOffset
 	}
-	s.dataReader.Reset(s.dataFd)
 
-	n, err := s.dataReader.Read(s.prologue)
-	if err != nil {
-		return nil, err
-	} else if n != cap(s.prologue) {
+	// We'll treat the dataView similar to a reader interface in that we will continue to reslice
+	// it such that the first byte is always the next byte, and we'll focus instead on how much
+	// we want to read.
+	dataView := s.dataMmap[entry.Offset:]
+
+	// Should never happen, but prevents panics in the case of malformed data
+	if len(dataView) < cap(s.prologue) {
 		return nil, errReadNotExpectedSize
-	} else if !bytes.Equal(s.prologue[:markerLen], marker) {
+	}
+	// Should never happen, but prevents panics in the case of malformed data
+	if !bytes.Equal(dataView[:markerLen], marker) {
 		return nil, errReadMarkerNotFound
 	}
 
-	var data checked.Bytes
+	// Reslice past the prologue that we just finished reading
+	dataView = dataView[cap(s.prologue):]
+
+	// Obtain an appropriately sized buffer
+	var buffer checked.Bytes
 	if s.bytesPool != nil {
-		data = s.bytesPool.Get(int(entry.Size))
-		data.IncRef()
-		defer data.DecRef()
-		data.Resize(int(entry.Size))
+		buffer = s.bytesPool.Get(int(entry.Size))
+		buffer.IncRef()
+		defer buffer.DecRef()
+		buffer.Resize(int(entry.Size))
 	} else {
-		data = checked.NewBytes(make([]byte, entry.Size), nil)
-		data.IncRef()
-		defer data.DecRef()
+		buffer = checked.NewBytes(make([]byte, entry.Size), nil)
+		buffer.IncRef()
+		defer buffer.DecRef()
 	}
 
-	n, err = s.dataReader.Read(data.Get())
-	if err != nil {
-		return nil, err
-	}
-
-	// In case the buffered reader only returns what's remaining in
-	// the buffer, repeatedly read what's left in the underlying reader.
-	for n < int(entry.Size) {
-		b := data.Get()[n:]
-		remainder, err := s.dataReader.Read(b)
-
-		if err == io.EOF {
-			break
-		} else if err != nil {
-			return nil, err
-		}
-		n += remainder
-	}
-
-	if n != int(entry.Size) {
-		return nil, errReadNotExpectedSize
-	}
+	// Copy the actual data into the underlying buffer
+	underlyingBuf := buffer.Get()
+	copy(underlyingBuf, dataView[:entry.Size])
```

Comment:
Should probably check `len(data) >= entry.Size` first yeah?

Reply:
Alternatively you can check this once before comparing against the marker, e.g. `if len(data) < markerLen + entry.Size { return nil, fmt.Errorf("not long enough...") }`